core/corehttp/metrics.go

```diff
@@ -29,6 +29,16 @@ func MetricsCollectionOption(handlerName string) ServeOption {
 			Objectives:  map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
 		}
 
+		if handlerName == "gateway" {
+			if err := prometheus.Register(unixfsGetMetric); err != nil {
+				if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+					unixfsGetMetric = are.ExistingCollector.(*prometheus.SummaryVec)
```

Comment:
The race detector is _not_ going to like that. Given how this is currently defined, we should probably just register this on init. We could, alternatively, store the metric in the gateway handler itself and register it on construction. But, as far as I can tell, init seems like the more correct approach (I have no idea, really).

Note: checking "gateway" doesn't really help given that the metric is a global anyways.

Reply:
The race detector doesn't _seem_ to care but it is out of place, I just didn't want a second place of metric registration. I am happy to move it to the gateway handler itself.

Reply:
Either way. We just shouldn't be setting a global in a function like this.

Reply:
Actually, would there ever be the case when that metric is already registered?

Reply:
Maybe but it shouldn't matter either way (I guess we could return an _error_ if it's already registered but we've registered the _wrong_ metric).